src/libraries/System.IO.Compression.Zstandard/src/System/IO/Compression/ZstandardStream.Compress.cs

```diff
@@ -0,0 +1,313 @@
+        /// <param name="stream">The stream to which compressed data is written or from which data to decompress is read.</param>
+        /// <param name="compressionLevel">One of the enumeration values that indicates whether to compress data to the stream or decompress data from the stream.</param>
+        /// <param name="leaveOpen"><see langword="true" /> to leave the stream open after the <see cref="ZstandardStream" /> object is disposed; otherwise, <see langword="false" />.</param>
+        public ZstandardStream(Stream stream, CompressionLevel compressionLevel, bool leaveOpen) : this(stream, CompressionMode.Compress, leaveOpen, new ZstandardEncoder(ZstandardUtils.GetQualityFromCompressionLevel(compressionLevel))) { }
+
+        /// <summary>Initializes a new instance of the <see cref="ZstandardStream" /> class by using the specified stream, options, and optionally leaves the stream open.</summary>
+        /// <param name="stream">The stream to which compressed data is written or from which data to decompress is read.</param>
+        /// <param name="compressionOptions">The options to use for Zstandard compression or decompression.</param>
+        /// <param name="leaveOpen"><see langword="true" /> to leave the stream open after the <see cref="ZstandardStream" /> object is disposed; otherwise, <see langword="false" />.</param>
+        public ZstandardStream(Stream stream, ZstandardCompressionOptions compressionOptions, bool leaveOpen = false) : this(stream, CompressionMode.Compress, leaveOpen, new ZstandardEncoder(compressionOptions)) { }
+
+        /// <summary>Initializes a new instance of the <see cref="ZstandardStream" /> class by using the specified stream and encoder instance.</summary>
+        /// <param name="stream">The stream to which compressed data is written.</param>
+        /// <param name="encoder">The encoder instance to use for compression. The stream will not dispose this encoder; instead, it will reset it when the stream is disposed.</param>
+        /// <param name="leaveOpen"><see langword="true" /> to leave the stream open after the <see cref="ZstandardStream" /> object is disposed; otherwise, <see langword="false" />.</param>
+        /// <exception cref="ArgumentNullException"><paramref name="stream"/> is <see langword="null"/>.</exception>
+        /// <exception cref="ArgumentException"><paramref name="stream"/> does not support writing.</exception>
+        public ZstandardStream(Stream stream, ZstandardEncoder encoder, bool leaveOpen = false) : this(stream, CompressionMode.Compress, leaveOpen, encoder)
+        {
+            _encoderOwned = false;
+        }
+
+        /// <summary>
+        /// Sets the size of the source data to be compressed.
+        /// </summary>
+        /// <param name="size">The size of the source data in bytes.</param>
+        /// <remarks>
+        /// This method must be called before writing data to the stream.
```

Comment:
Interesting, if this always must be set for compression, should we make it a constructor parameter?  Requiring up-front knowledge of size seems to limit possible streaming scenarios.

Reply:
In that case please improve the remarks so that they don't use "must", or qualify it to make the consequences clear.  Describe what happens if set after writing data, or if never setting at all.